test/helpers/sockets.rb

```diff
@@ -0,0 +1,435 @@
+          raise e unless IO.select(nil, [io], nil, 5)
+          retry
+        rescue Errno::EPIPE, SystemCallError, IOError => e
+          raise e
+        end
+
+        return if n == str.bytesize
+        str = str.byteslice(n..-1)
+      end
+    end
+
+    private
+
+    def get_req(path = nil, dly: nil, len: nil)
+      req = "GET /#{path} HTTP/1.1\r\n".dup
+      req << "Dly: #{dly}\r\n" if dly
+      req << "Len: #{len}\r\n" if len
+      req << "\r\n"
+    end
+  end
+
+  class SktTCP < ::TCPSocket
+    prepend SktPrepend
+  end
+
+  if Object.const_defined? :UNIXSocket
+    class SktUnix < ::UNIXSocket
+      prepend SktPrepend
+    end
+  end
+
+  if !Object.const_defined?(:Puma) || ::Puma.ssl?
+    require 'openssl'
+    class SktSSL < ::OpenSSL::SSL::SSLSocket
+      prepend SktPrepend
+    end
+  end
+
+  module Sockets
+
+    IS_JRUBY = Object.const_defined? :JRUBY_VERSION
+
+    IS_OSX = RUBY_PLATFORM.include? 'darwin'
+
+    IS_WINDOWS = !!(RUBY_PLATFORM =~ /mswin|ming|cygwin/ ||
+      IS_JRUBY && RUBY_DESCRIPTION.include?('mswin'))
+
+    IS_MRI = (RUBY_ENGINE == 'ruby' || RUBY_ENGINE.nil?)
+
+    HOST = TestPuma.const_defined?(:SvrBase) ? SvrBase::HOST : '127.0.0.1'
+
+    # OpenSSL::SSL::SSLError is intermittently raised on SSL connect?
+    # IOError is intermittent on all platforms (except maybe Windows), seems
+    # to be only be raised on SSL connect?
+    # macOS sometimes raises Errno::EBADF for socket init error
+    #
+    OPEN_WRITE_ERRORS = begin
+      ary = [Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::ENOENT,
+        Errno::EPIPE, Errno::EBADF, IOError]
+      ary << Errno::ENOTCONN if IS_OSX
+      ary << OpenSSL::SSL::SSLError if Object.const_defined?(:OpenSSL)
+
+      ary.freeze
+    end
+
+
+    def fast_connect_get_body(path = nil, dly: nil, len: nil, timeout: READ_TIMEOUT)
+      fast_connect_get(path, dly: dly, len: len).read_body timeout
+    end
+
+    def fast_connect_get_response(path = nil, dly: nil, len: nil, timeout: READ_TIMEOUT)
+      fast_connect_get(path, dly: dly, len: len).read_response timeout
+    end
+
+    # use only if all socket writes are fast
+    # does not wait for a read
+    def fast_connect_get(path = nil, dly: nil, len: nil)
+      req = "GET /#{path} HTTP/1.1\r\n".dup
+      req << "Dly: #{dly}\r\n" if dly
+      req << "Len: #{len}\r\n" if len
+      req << "\r\n"
+      fast_connect_raw req
+    end
+
+    # Use to send a raw string, keyword parameters are the same as `fast_connect`.
+    #
+    def fast_connect_raw(str = nil, type: nil, p: nil)
+      s = fast_connect type: type, p: p
+      s.fast_write str
+      s
+    end
+
+    # Used to open a socket.  Normally, the connection info is supplied by calling `bind_type`,
+    # but that can be set manually if needed for control sockets, etc.
+    # @param type: [Symbol] the type of connection, eg, :tcp, :ssl, :aunix, :unix
+    # @param p: [String, Integer] the port or path of the connection
```

Comment:
`p` -> `file_descriptor` perhaps?

Reply:
maybe even just `fd`, we do have precedent for using `fd` elsewhere

Reply:
Actually that is either the TCP/SSL port or the UNIXSocket path.  Not sure what to call it, but the doc/comments could be better.

Reply:
maybe `socket`, because that's what both of those things are then

Reply:
Maybe `port_path`.  Some might think `socket` is a socket instance?